----
# **<span style="color:darkmagenta">PROYECTO EDA - BANK MARKETING</span>**
# **<span style="color:darkmagenta">Análisis preliminar</span>**
----

---
---
## <span style="color:gray">**1. Importación de librerías**</span> 📂

In [1]:
# Tratamiento de datos
import numpy as np
import pandas as pd 

# Configuración de ruta
import sys
sys.path.append('../')

# Importación de funciones personalizadas
from src.soporte1_preliminar import eda_preliminar

---
---
## <span style="color:gray">**2. Carga de datos**</span> 📥

In [2]:
# Datos CSV
bank_data = pd.read_csv("../data/raw/bank-additional.csv")

# Datos Excel
customer_data = pd.read_excel("../data/raw/customer-details.xlsx", sheet_name=None, index_col=0) # Devuelve un diccionario con todas las hojas

In [3]:
# Unión de las hojas de excel en un solo DataFrame
customer_data = pd.concat(customer_data.values(), ignore_index=True)

In [4]:
# Configuración para mostrar todas las columnas
pd.set_option('display.max_columns', None)

---
---


----
# <span style="color:darkmagenta">**Desarrollo del proyecto - 1**</span>
----

---
---
## <span style="color:gray">**Exploración inicial de los datasets**</span> 🔎

Antes de comenzar el análisis exploratorio de datos (EDA), es importante realizar una inspección general de los datasets. Este primer contacto con los datos nos permite entender su estructura, el tipo de variables que contienen y detectar posibles problemas como valores faltantes o formatos incorrectos.

Para ello utilizamos varias funciones de Pandas que nos ayudan a obtener una visión rápida y completa del conjunto de datos.

---

## `bank_data`

In [5]:
# Llamamos a la función eda_preliminar guardada en el archivo de sorporte
eda_preliminar(bank_data)



👀 Visualización de las primeras y últimas filas



,Unnamed: 0,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
0,0,NaN,housemaid,MARRIED,basic.4y,0.0,0.0,0.0,telephone,261,1,999,0,NONEXISTENT,1.1,"93,994","-36,4","4,857",5191,no,2-agosto-2019,41.495,-71.233,089b39d8-e4d0-461b-87d4-814d71e0e079
1,1,57.0,services,MARRIED,high.school,NaN,0.0,0.0,telephone,149,1,999,0,NONEXISTENT,1.1,"93,994","-36,4",NaN,5191,no,14-septiembre-2016,34.601,-83.923,e9d37224-cb6f-4942-98d7-46672963d097
2,2,37.0,services,MARRIED,high.school,0.0,1.0,0.0,telephone,226,1,999,0,NONEXISTENT,1.1,"93,994","-36,4","4,857",5191,no,15-febrero-2019,34.939,-94.847,3f9f49b5-e410-4948-bf6e-f9244f04918b


,Unnamed: 0,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
42997,15046,NaN,blue-collar,SINGLE,basic.6y,0.0,1.0,0.0,cellular,391,2,999,0,NONEXISTENT,1.4,"93,918","-42,7",NaN,"5228,1",no,15-septiembre-2016,40.679,-120.015,cadadd4b-7ee5-4019-b13a-ca01bb67ca5b
42998,15280,NaN,admin.,MARRIED,university.degree,NaN,0.0,0.0,cellular,674,3,999,0,NONEXISTENT,1.4,"93,918","-42,7","4,958","5228,1",no,23-septiembre-2019,27.772,-117.518,5f432048-d515-4bb5-9c94-62db451f88d4
42999,27570,NaN,unemployed,SINGLE,university.degree,0.0,0.0,1.0,cellular,104,2,999,0,NONEXISTENT,-0.1,"93,2",-42,"4,021","5195,8",no,6-noviembre-2019,41.146,-105.026,993bbbd6-4dbc-4a40-a408-f91f8462bee6




------------------------------------------------------------------------


🎲 Visualización de filas aleatorias



,Unnamed: 0,age,job,marital,education,default,housing,loan,contact,duration,campaign,pdays,previous,poutcome,emp.var.rate,cons.price.idx,cons.conf.idx,euribor3m,nr.employed,y,date,latitude,longitude,id_
23703,23800,50.0,blue-collar,MARRIED,basic.4y,NaN,1.0,0.0,cellular,73,1,999,0,NONEXISTENT,1.4,"93,444","-36,1","4,963","5228,1",no,26-abril-2019,38.169,-67.020,247eb5a8-e4b7-4d68-9e81-2e0a0a382b8c
32313,32457,32.0,blue-collar,MARRIED,basic.9y,NaN,1.0,0.0,cellular,177,2,999,0,NONEXISTENT,-1.8,"92,893","-46,2","1,313","5099,1",no,20-febrero-2019,28.756,-72.464,90c52c13-3b98-4d32-a527-c7dd41e194d7
33199,33343,28.0,unemployed,MARRIED,high.school,NaN,0.0,0.0,cellular,216,1,999,1,FAILURE,-1.8,"92,893","-46,2",NaN,"5099,1",no,19-octubre-2015,42.129,-82.200,19cfae34-b428-408b-9a40-e0eae96dbafc




------------------------------------------------------------------------


📏 Dimensión del DataFrame

Número de filas del DataFrame: 43000
Número de columnas del DataFrame: 24


------------------------------------------------------------------------


🧾 Información general

<class 'pandas.DataFrame'>
RangeIndex: 43000 entries, 0 to 42999
Data columns (total 24 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      43000 non-null  int64  
 1   age             37880 non-null  float64
 2   job             42655 non-null  str    
 3   marital         42915 non-null  str    
 4   education       41193 non-null  str    
 5   default         34019 non-null  float64
 6   housing         41974 non-null  float64
 7   loan            41974 non-null  float64
 8   contact         43000 non-null  str    
 9   duration        43000 non-null  int64  
 10  campaign        43000 non-null  int64  
 11  pdays           43000 non-null  in

Index(['Unnamed: 0', 'age', 'job', 'marital', 'education', 'default',
       'housing', 'loan', 'contact', 'duration', 'campaign', 'pdays',
       'previous', 'poutcome', 'emp.var.rate', 'cons.price.idx',
       'cons.conf.idx', 'euribor3m', 'nr.employed', 'y', 'date', 'latitude',
       'longitude', 'id_'],
      dtype='str')



------------------------------------------------------------------------


🔢 Estadísticas descriptivas de las variables numéricas



,count,mean,std,min,25%,50%,75%,max
Unnamed: 0,43000.0,20651.10,11868.77,0.0,10381.75,20642.50,30930.25,41187.00
age,37880.0,39.98,10.44,17.0,32.00,38.00,47.00,98.00
default,34019.0,0.00,0.01,0.0,0.00,0.00,0.00,1.00
housing,41974.0,0.54,0.50,0.0,0.00,1.00,1.00,1.00
loan,41974.0,0.16,0.36,0.0,0.00,0.00,0.00,1.00
duration,43000.0,257.74,258.67,0.0,102.00,179.00,319.00,4918.00
campaign,43000.0,2.57,2.77,1.0,1.00,2.00,3.00,56.00
pdays,43000.0,962.33,187.26,0.0,999.00,999.00,999.00,999.00
previous,43000.0,0.17,0.50,0.0,0.00,0.00,0.00,7.00
emp.var.rate,43000.0,0.08,1.57,-3.4,-1.80,1.10,1.40,1.40




------------------------------------------------------------------------


🔡 Estadísticas descriptivas de las variables categóricas



,count,unique,top,freq
job,42655,11,admin.,10873
marital,42915,3,MARRIED,25999
education,41193,7,university.degree,12722
contact,43000,2,cellular,27396
poutcome,43000,3,NONEXISTENT,37103
cons.price.idx,42529,26,"93,994",7938
cons.conf.idx,43000,26,"-36,4",8020
euribor3m,33744,309,"4,857",2287
nr.employed,43000,11,"5228,1",16980
y,43000,2,no,38156


### **Informe de EDA preliminar – Dataset `bank_data`**

**1. Información general del Dataset**

- El Dataset contiene 43000 registros.

- Incluye columnas numéricas, categóricas y temporales.

- La columna `Unnamed: 0` parece un índice automático, con valores que no coinciden con el número de filas, sugiriendo posibles duplicados o nulos.

- Se identifican inconsistencias de formato, nombres de columnas poco claros y caracteres innecesarios.

- Se identifican valores nulos en varias columnas, inconsistencias en mayúsculas/minúsculas y signos de puntuación incorrectos.

**2. Hallazgos por tipo de variable**

Variables numéricas:

- `age`: media aproximadamente 40 años, rango 17–98.

- `duration`: presenta alta dispersión y posibles outliers.

- `campaign` y `previous`: la mayoría de los clientes han sido contactados pocas veces; previous parece ser siempre 0.

- `pdays`: concentración de valores en 999, indicando clientes no contactados previamente.

Variables categóricas:

- `job`, `marital`, `education`: múltiples categorías, algunas muy concentradas en ciertos valores.

- `contact`: dominado por *cellular*.

- `poutcome`: mayoría en *NONEXISTENT*, indicando baja participación en campañas previas.

Variables como `cons.price.idx`, `cons.conf.idx`, `euribor3m` y `nr.employed` están siendo tratadas como categóricas y deberán convertirse a numéricas.

Variable objetivo:

- `y` está fuertemente desbalanceada, con predominio de *no*.

Variables temporales

- `date`: aparece como variable categórica, por lo que habría que transformarla. También se observa que contiene un gran número de valores únicos, lo que indica alta variabilidad temporal y la necesidad de transformarla en variables más informativas (mes, año).


💡 **Conclusión:**

El Dataset requiere limpieza y estandarización antes del análisis, debido a valores nulos, inconsistencias y desbalances. Abordar estas cuestiones permitirá generar insights precisos y confiables.

---
## `customer_data`

In [ ]:
# Llamamos a la función eda_preliminar guardada en el archivo de sorporte
eda_preliminar(customer_data)



👀 Visualización de las primeras y últimas filas



,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID
0,161770,1,0,2012-04-04,29,089b39d8-e4d0-461b-87d4-814d71e0e079
1,85477,1,1,2012-12-30,7,e9d37224-cb6f-4942-98d7-46672963d097
2,147233,1,1,2012-02-02,5,3f9f49b5-e410-4948-bf6e-f9244f04918b


,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID
43167,78470,2,2,2014-10-28,12,d6271666-319d-42c8-a741-cb22bf2c2093
43168,144218,0,1,2014-07-01,9,f67fbfa8-6573-414d-a805-b26a2f1b1ceb
43169,115155,0,2,2014-12-25,7,9727bc06-c11a-461a-a5bb-3d210467cc2a




------------------------------------------------------------------------


🎲 Visualización de filas aleatorias



,Income,Kidhome,Teenhome,Dt_Customer,NumWebVisitsMonth,ID
20986,48186,1,0,2013-06-12,3,ebab348d-7dbe-49cb-a89c-1dfca1ce9264
27623,90535,2,2,2013-01-05,30,f2f02abe-be8b-4faf-812e-fafee4f0f02e
30028,130241,2,1,2014-01-05,7,b554f8a3-9a45-41cd-b249-c25a6aa11acc




------------------------------------------------------------------------


📏 Dimensión del DataFrame

Número de filas del DataFrame: 43170
Número de columnas del DataFrame: 6


------------------------------------------------------------------------


🧾 Información general

<class 'pandas.DataFrame'>
RangeIndex: 43170 entries, 0 to 43169
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Income             43170 non-null  int64         
 1   Kidhome            43170 non-null  int64         
 2   Teenhome           43170 non-null  int64         
 3   Dt_Customer        43170 non-null  datetime64[us]
 4   NumWebVisitsMonth  43170 non-null  int64         
 5   ID                 43170 non-null  str           
dtypes: datetime64[us](1), int64(4), str(1)
memory usage: 2.0 MB


------------------------------------------------------------------------


❌ Detección de nulos

No hay nulos


-------

Index(['Income', 'Kidhome', 'Teenhome', 'Dt_Customer', 'NumWebVisitsMonth',
       'ID'],
      dtype='str')



------------------------------------------------------------------------


🔢 Estadísticas descriptivas de las variables numéricas



,count,mean,min,25%,50%,75%,max,std
Income,43170.0,93227.39,5841.0,49608.0,93009.5,136740.5,180802.0,50498.18
Kidhome,43170.0,1.0,0.0,0.0,1.0,2.0,2.0,0.82
Teenhome,43170.0,1.0,0.0,0.0,1.0,2.0,2.0,0.82
Dt_Customer,43170,2013-05-11 22:09:47.407922,2012-01-01 00:00:00,2012-07-16 00:00:00,2013-03-01 00:00:00,2014-03-28 00:00:00,2014-12-31 00:00:00,NaN
NumWebVisitsMonth,43170.0,16.59,1.0,9.0,17.0,25.0,32.0,9.24




------------------------------------------------------------------------


🔡 Estadísticas descriptivas de las variables categóricas



,count,unique,top,freq
ID,43170,43170,089b39d8-e4d0-461b-87d4-814d71e0e079,1


### **Informe de EDA preliminar – Dataset `customer_data`**

**1. Información general del Dataset**

- El Dataset contiene 43170 registros y 6 variables.

- No se identifican valores nulos ni registros duplicados.

- La estructura es consistente y no presenta problemas evidentes de formato o calidad inicial de datos.

**2. Hallazgos por tipo de variable**

Variables numéricas:

- `Income`: presenta una alta variabilidad, con valores entre 5841 y 180802, lo que indica diferencias significativas en el nivel de ingresos y posibles valores extremos.

- `NumWebVisitsMonth`: media de aproximadamente 16 visitas mensuales, con un rango entre 1 y 32, reflejando distintos niveles de interacción digital entre los clientes.

Variables de composición del hogar:

`Kidhome` y `Teenhome`: toman valores entre 0 y 2, con medias cercanas a 1, lo que sugiere que la mayoría de los clientes tiene uno o ningún hijo en casa.

💡 **Conclusión:**

El Dataset presenta una estructura limpia y consistente, sin valores nulos ni duplicados, lo que facilita su análisis. Sin embargo, la alta variabilidad en variables como `Income` sugiere la necesidad de evaluar posibles outliers. Además, se realizará un proceso de renombrado de variables para mejorar la claridad del Dataset. En general, los datos son adecuados para el análisis y permitirán obtener insights relevantes sobre el comportamiento y perfil de los clientes.